### DISCUSSION TITLE
Structured concurrency solutions

### DISCUSSION THREAD

**Discussion by cl4cnam** [cl4cnam/funcSug#25] [Ideas]
Just to feed the thought about the [article of Francisco Sant'Anna](https://fsantanna.github.io/structured-concurrency.html), here are
some of what I feel to contain (at least partial) "structured concurrency" solutions (classified by base language or target language):

- C: [Céu](https://github.com/ceu-lang/ceu) (by Francisco Sant'Anna), [Esterel](http://www-sop.inria.fr/meije/esterel/esterel-eng.html) (by Gérard Berry), [libdill](https://github.com/sustrik/libdill) (by Martin Sustrik), Reactive C (by Frédéric Boussinot), Reactive Scripts (by Frédéric Boussinot), SL (by Frédéric Boussinot)
- C++: [Senders](https://github.com/nvidia/stdexec)
- C#: [StructuredConcurrency](https://github.com/StephenCleary/StructuredConcurrency) (by Stephen Cleary)
- Crystal: [nested_scheduler](https://github.com/yxhuvud/nested_scheduler) (by Linus Sellberg)
- Dart: [structured_async](https://github.com/renatoathaydes/structured_async) (by Renato Athaydes)
- F#: [Blech](https://www.blech-lang.org/) (by Franz-Josef Grosch)
- Go: [conc](https://github.com/sourcegraph/conc) (by Camden Cheek), [Flowmatic](https://github.com/earthboundkid/flowmatic) (by Carlana), [nursery](https://github.com/arunsworld/nursery)
- Haskell: [Control.Concurrent.Async](https://hackage.haskell.org/package/async-2.2.5/docs/Control-Concurrent-Async.html) (by Simon Marlow), [ki](https://github.com/awkward-squad/ki) (by Mitchell Rosen)
- Hylo: [Hylo](https://docs.hylo-lang.org/language-tour/concurrency) [2024-01-08](https://github.com/hylo-lang/hylo/pull/1283)
- Java or JVM: Icobjs (by Frédéric Boussinot), [jcsp](https://www.cs.kent.ac.uk/projects/ofa/jcsp/) (by Paul Austin), [JDK 20](https://techspoti.com/jdk-20-the-brand-new-options-in-java-20/), [Jox](https://github.com/softwaremill/jox) (by Adam Warski), Rejo (by Raul Acosta-Bermejo), SugarCubes (by Jean-Ferdy Susini)
- JavaScript: [Epistrophy](https://github.com/stcucufa/epistrophy) (by Ryoji Ikea), [FuncSug](https://github.com/cl4cnam/funcSug) (that's mine), [nursery](https://github.com/giltayar/nursery) (by Gil Tayar), [QED](https://qed-lang.org) (by Martin Savage), [StratifiedJs](https://github.com/onilabs/stratifiedjs), [SugarCubesJS](https://github.com/LordManta/SugarCubesJS) (by Jean-Ferdy Susini)
- Julia: [Julio](https://github.com/JuliaConcurrent/Julio.jl) (by Takafumi Arakaki)
- Kotlin: [Céu](https://github.com/fsantanna/dceu) (by Francisco Sant'Anna), [kotlinx.coroutines](https://github.com/Kotlin/kotlinx.coroutines) (by Roman Elizarov)
- Lua: [Atmos](https://github.com/lua-atmos/atmos) (by Francisco Sant'Anna)
- ML: [Reactive ML](https://github.com/reactiveml/rml) (by Louis Mandel)
- OCaml: [Pendulum](https://github.com/remyzorg/pendulum) (by Rémy El Sibaïe Besognet)
- Python: [AnyIO](https://github.com/agronholm/anyio) (by Alex Grönholm), [Curio](https://github.com/dabeaz/curio) (by David Beazley), Python (starting from Python 3.11), [quattro](https://github.com/Tinche/quattro) (by Tin Tvrtković), [Trio](https://github.com/python-trio/trio) (by Nathaniel J. Smith), [trio-util](https://github.com/groove-x/trio-util) (by John Belmonte)
- Ruby: [Polyphony](https://github.com/digital-fabric/polyphony) (by Sharon Rosner)
- Rust: [async_nursery](https://github.com/najamelan/async_nursery) (by Naja Melan)
- Scala: [Ox](https://github.com/softwaremill/ox) (by Adam Warski)
- Swift: [Pappe](https://github.com/frameworklabs/Pappe) (by Framework Labs), [Swift 5.5](https://www.hackingwithswift.com/swift/5.5/structured-concurrency), [Venice](https://github.com/Zewo/Venice) (by Paulo Faria)
- TypeScript: [effection](https://github.com/thefrontside/effection)

Note:

- These are only about the quadrant C of [his diagram](https://fsantanna.github.io/structured-concurrency.html).
- I've included imperative synchronous languages into these.

Without any guarantee. Don't hesitate to fix me!

**Comment by Brian-M-J** [comment]
There are also [C++](https://github.com/nvidia/stdexec)'s and [Hylo](https://youtu.be/ws-Z8xKbP4w?t=2752)'s senders and receivers. I don't know which quadrant they would fall into though.

**Reply by cl4cnam** [reply]
Thank you very much for your contribution. :)

Yes, I think you're right. So, I include your two examples.

Can you tell me from which version Hylo includes structured concurrency?

**Reply by Brian-M-J** [reply]
>Can you tell me from which version Hylo includes structured concurrency?

Hylo is still in a very early stage; I couldn't find any version numbers on their [GitHub](https://github.com/hylo-lang/hylo/releases). All I could find was this [pull request](https://github.com/hylo-lang/hylo/pull/1283) to add concurrency support to Hylo.

**Reply by cl4cnam** [reply]
Ok, thanks. :slightly_smiling_face:

**Comment by Brian-M-J** [comment]
There are also tons of libraries and frameworks for the actor model (implemented by Erlang) and process calculi:
- [CAF](https://www.actor-framework.org/)
- [Apache Pekko](https://pekko.apache.org/)
- [Actix](https://actix.rs/docs/actix/actor/)
- the [stlab](https://stlab.cc/libraries/concurrency/channel.hpp/) concurrency library
- [crimson-rust](https://github.com/sinclairzx81/crimson-rust)
- [CSPLib](https://github.com/akiradeveloper/csplib)
- [Orleans](https://github.com/dotnet/orleans)
- [Pony](https://www.ponylang.io/)

And in FRP:
- [Rx](https://reactivex.io/languages.html)
- [reactive_rs](https://docs.rs/reactive-rs/latest/reactive_rs/)

Naming _all_ the concurrency solutions in existence right now seems too daunting. There's been a ton of work in this area.

**Reply by cl4cnam** [reply]
I think you're very right :slightly_smiling_face: but I don't want to name all concurrency solutions, only _[structured](https://en.wikipedia.org/wiki/Structured_concurrency)_ ones. :wink:

**Reply by Brian-M-J** [reply]
I'd say that unstructured concurrency libraries don't solve concurrency in any way whatsoever.
